packages/flutter/lib/src/painting/text_painter.dart

```diff
@@ -636,7 +640,7 @@ class TextPainter {
       return null;
 
     // Check for multi-code-unit glyphs such as emojis or zero width joiner
-    final bool needsSearch = _isUtf16Surrogate(prevCodeUnit) || _text.codeUnitAt(offset) == _zwjUtf16;
+    final bool needsSearch = _isUtf16Surrogate(prevCodeUnit) || _text.codeUnitAt(offset) == _zwjUtf16 || _isUnicodeDirectionality(prevCodeUnit);
```

Comment:
Directionality markers take up zero space. This makes sure the range covers at least one real glyph when computing carets when it encounters a directionality marker